lib/cloud/gcp/vm.go

```diff
@@ -0,0 +1,545 @@
+	}
+	if err := op.Wait(ctx); err != nil {
+		return trace.Wrap(err)
+	}
+	return nil
+}
+
+// RunCommandRequest contains parameters for running a command on an instance.
+type RunCommandRequest struct {
+	// Client is the instance client to use.
+	Client InstancesClient
+	// InstanceRequest is the set of parameters identifying the instance.
+	InstanceRequest
+	// Script is the script to execute.
+	Script string
+	// SSHPort is the ssh server port to connect to. Defaults to 22.
+	SSHPort     string
+	dialContext func(ctx context.Context, network, addr string) (net.Conn, error)
+}
+
+func (req *RunCommandRequest) CheckAndSetDefaults() error {
+	if err := req.InstanceRequest.CheckAndSetDefaults(); err != nil {
+		return trace.Wrap(err)
+	}
+	if req.Script == "" {
+		return trace.BadParameter("script must be set")
+	}
+	if req.SSHPort == "" {
+		req.SSHPort = "22"
+	}
+	if req.dialContext == nil {
+		dialer := net.Dialer{}
+		req.dialContext = dialer.DialContext
+	}
+	return nil
+}
+
+func generateKeyPair() (ssh.Signer, ssh.PublicKey, error) {
+	rawPriv, rawPub, err := native.GenerateKeyPair()
+	if err != nil {
+		return nil, nil, trace.Wrap(err)
+	}
+	signer, err := ssh.ParsePrivateKey(rawPriv)
+	if err != nil {
+		return nil, nil, trace.Wrap(err)
+	}
+	publicKey, _, _, _, err := ssh.ParseAuthorizedKey(rawPub)
+	if err != nil {
+		return nil, nil, trace.Wrap(err)
+	}
+	return signer, publicKey, nil
+}
+
+// RunCommand runs a command on an instance.
+func RunCommand(ctx context.Context, req *RunCommandRequest) error {
+	if err := req.CheckAndSetDefaults(); err != nil {
+		return trace.Wrap(err)
+	}
+
+	// Generate keys and add them to the instance.
+	signer, publicKey, err := generateKeyPair()
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	instance, err := req.Client.GetInstance(ctx, &req.InstanceRequest)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	if len(instance.hostKeys) == 0 {
+		return trace.NotFound(`Instances %v is missing host keys. Did you enable guest attributes on the instance?
+https://cloud.google.com/solutions/connecting-securely#storing_host_keys_by_enabling_guest_attributes`, req.Name)
+	}
+	if instance.ipAddress == "" {
+		return trace.NotFound("Instance %v is missing a public ip address.", req.Name)
+	}
+	keyReq := &SSHKeyRequest{
+		Instance:  instance,
+		PublicKey: publicKey,
+	}
+	if err := req.Client.AddSSHKey(ctx, keyReq); err != nil {
+		return trace.Wrap(err)
+	}
+
+	// Clean up the key when we're done.
+	defer func() {
+		var err error
+		// Fetch the instance first to get the most up-to-date metadata hash.
+		if keyReq.Instance, err = req.Client.GetInstance(ctx, &req.InstanceRequest); err != nil {
+			logrus.WithError(err).Warn("Error fetching instance.")
+			return
+		}
+		if err := req.Client.RemoveSSHKey(ctx, keyReq); err != nil {
+			logrus.WithError(err).Warn("Error deleting SSH Key.")
+		}
+	}()
+
+	// Configure the SSH client.
+	callback, err := sshutils.HostKeyCallback(instance.hostKeys, true)
```

Comment:
Any way to factor the code from here onwards into a some kind of "RunSSH" helper method, maybe put it in sshutils?